### DISCUSSION TITLE
impossible to add my erc20 token to metamask

### DISCUSSION THREAD

**Discussion by ltixier2** [ConsenSys-archive/ganache#4332] [General]
i have an hosted ganache-cli on a linux server. 
I have deployed the ERC20 contract with remix. 
when i try to set the contract address as a new token in metamask's assets, the token is not shown.

**Comment by jeffsmale90** [comment]
Hi @ltixier2 I've converted this to a discussion as this is a better place to help you with your issue, and helps other people facing the same challenge.

This certainly should work, and I've reproduced myself locally. can you please share some more details:

- Can you check the balance directly via the `balanceOf` ERC20 method on the deployed contract 
- When you import the token into Metamask, are the details (such as Symbol, Decimals) populating?
- Check the logs of Ganache to see whether Metamask is requesting the details (there should be a number of `eth_call`s when you enter the contract address into the `Import token` dialog
- Is the correct Eth balance showing in Metamask for the Ganache account

**Comment by bztrn** [comment]
Hi @jeffsmale90 

I'm facing the same issue and sharing the details that you've asked for:

1) Yes.
2) Yes. (Inc. Balance of the token)
3) 
```

eth_call: [{
    "to": "0xD0b1c0c2Abad7EEB7B3F6D396FB47bb4b5257cd0",
    "data": "0x313ce567"
    },
    "0x2"
    ]
    
eth_call: [{
    "to": "0xD0b1c0c2Abad7EEB7B3F6D396FB47bb4b5257cd0",
    "data": "0x70a0823100000000000000000000000049ac411f9c67a9008f856555beffd3b64fe84c6d"
    },
    "0x2"
    ]
    
eth_call: [{
    "to": "0xd0b1c0c2abad7eeb7b3f6d396fb47bb4b5257cd0",
    "data": "0x01ffc9a780ac58cd00000000000000000000000000000000000000000000000000000000"
    },
    "0x2"
    ]

```
4) Yes.

Some Additional Informations:
- OS: macOS Monterey 12.6
- Metamask: on Google Chrome 111.0.5563.64
- Installed Version of Ganache: v2.7.0
- Configuration: Hostname - 0.0.0.0 All Interfaces | Port - 7545 | Network ID - 1337
- Importing custom token on internal wallet of the Brave browser is done successfully. It is showing up the token and its balance on the wallet, but, I am unable to send / make token transfers since it is not displaying the token on the transfer screen.

**Reply by jeffsmale90** [reply]
Thanks for the info! I've not tried with Ganache UI, but (assuming it's a new workspace, rather than a legacy one - does it have a ⚠️ icon beside it's name in the workspace list?) it will be using ganache v7 under the hood, so it shouldn't be a problem.

I'm able to transfer the ERC20 with metamask, but I haven't tried with Brave - I'll try that out tomorrow and let you know how I get on.

**Reply by bztrn** [reply]
Thanks for the answer.

Yes, it's a new workspace and it is doesn't have a ⚠️ icon beside it's name in the list.

> I'm able to transfer the ERC20 with metamask, but I haven't tried with Brave - I'll try that out tomorrow and let you know how I get on.

I am looking forward to hearing from you regarding that as well

**Comment by ltixier2** [comment]
hello : 



> Can you check the balance directly via the balanceOf ERC20 method on the deployed contract? 
 
Yes it works 



> When you import the token into Metamask, are the details (such as Symbol, Decimals) populating?sure when importing it shows the token name and the balance. Once i click in import the token is not in the list. 

Once i click on the import llink, the token with correct decimals is shown, also the balance. Once it is imported it is not added into the list of tokens. 


> Check the logs of Ganache to see whether Metamask is requesting the details (there should be a number of eth_calls when you enter the contract address into the Import token dialog

Yes some eth_calls done. 


> Is the correct Eth balance showing in Metamask for the Ganache account
balance is ok. 


The same procedure had been tested with hardhat, no issue. 


Ganache CLI v6.12.2 (ganache-core: 2.13.2)


ganache-cli -d --db ./ganacheDatas --accounts 20  --host 0.0.0.0
(mnemonic is hidden)

**Reply by jeffsmale90** [reply]
@ltixier2 thanks for the info - the bump from v6 to v7 was a significant re-work of the codebase (complete rewrite), so this may be an issue with the older version. Are you able to upgrade to the latest?

**Comment by ltixier2** [comment]
@jeffsmale90 
after upgrading : 
ganache v7.7.7 (@ganache/cli: 0.8.6, @ganache/core: 0.8.6)

exactly the same issue

**Reply by jeffsmale90** [reply]
Ok, good to know. I'll pick this up again tomorrow and see if I can figure out what is going on. 

Thanks!

**Comment by jeffsmale90** [comment]
@bztrn I have managed to reproduce the issue with Brave browser, where I can transfer the token with Hardhat network, but not with Ganache.

@ltixier2 I have not yet managed to reproduce the issue where the token does not populate in MetaMask.

I've reached out to the MetaMask team for feedback on how this could happen, and I'm digging more into the Brave browser issue. they may be caused by the same underlying issue 🤷

**Comment by jeffsmale90** [comment]
@bztrn we've figured out what's going on with Brave wallet - it's conflating `networkId` and `chainId`. In order for the token to appear in the swap list, `networkId` and `chainId` must be the same.

For historical reasons, Ganache creates a random `networkId` on startup.

You can work around this with `ganache --chain.networkId 1337` which matches the default `chainId`.

We have also created an issue to look into retiring this behaviour: #4334

@ltixier2 I've found a behaviour in MetaMask where it is attempting to fetch the token details using an explicit block number (rather than `latest`). It's possible for MetaMask and Ganache to become out of sync, and MetaMask will never reduce it's block number. One way to work around this is to reset your account in MetaMask - this will resync block number and other data that is stored within MetaMask. https://support.metamask.io/hc/en-us/articles/360015488891-How-to-reset-an-account

I've also raised with the MetaMask team this issue, to see whether they can improve the behaviour.

Please let me know if this helps in both cases! Otherwise I'll keep digging.

**Reply by jeffsmale90** [reply]
Thanks to @davidmurdoch of helping debug this!

**Reply by ltixier2** [reply]
sorry but after adding --chain.networkId 1337  and reseting the metamask account. The issue is still the same.

**Reply by jeffsmale90** [reply]
Can you please try changing your chainId to 1?

**Comment by ltixier2** [comment]
Hello, 
i confirm that the chainId 1 is working

**Reply by davidmurdoch** [reply]
Are you setting both chainId and networkId to 1?

**Reply by ltixier2** [reply]
--host 0.0.0.0 --chain.networkId 1 --chain.chainId 1

**Reply by davidmurdoch** [reply]
Did you also try setting both networkId and chainId to 1337?

**Reply by jeffsmale90** [reply]
Spoke with someone from the metamask team, and understand that the 1337 chainId s not supported by the erc20 logic. I'll work with them to figure out how to get this resolved. 

In the meantime, I'm glad we could get a workaround that hopefully unblocks you!

Please let us know if you have any more problems.
